File: dynamic_programming/01_Fibonacci/01_Fibonacci.js
const testCases = [
  { n: 0, expect: 0 },
  { n: 1, expect: 1 },
  { n: 2, expect: 1 },
  { n: 3, expect: 2 },
  { n: 4, expect: 3 },
  { n: 5, expect: 5 },
  { n: 10, expect: 55 },
  { n: 20, expect: 6765 },
  { n: 30, expect: 832040 },
];

const fibFunctions = {
  Recursive: fibRecursive,
  Memoization: fibMemoization,
  Tabulation: fibTabulation,
  SpaceOptimized: fibSpaceOptimized,
};

function runTests() {
  // Encapsulate tests in a function
  for (const name in fibFunctions) {
    console.log(`Testing: ${name}`);
    testCases.forEach((testCase, i) => {
      const result = fibFunctions[name](testCase.n);
      console.log(
        `Test Case ${i + 1}: fib(${testCase.n}) = ${result}, Expected: ${
          testCase.expect
        }`
      );
      if (result === testCase.expect) {
        console.log("PASS");
      } else {
        console.log("FAIL");
      }
    });
    console.log();
  }
}

function main() {
  const n = 10;
  console.log(`fibRecursive(${n}) = ${fibRecursive(n)}`);
  console.log(`fibMemoization(${n}) = ${fibMemoization(n)}`);
  console.log(`fibTabulation(${n}) = ${fibTabulation(n)}`);
  console.log(`fibSpaceOptimized(${n}) = ${fibSpaceOptimized(n)}`);
  runTests(); // Call the tests
}

main(); // Call the main function

// Recursive approach
function fibRecursive(n) {
  if (n <= 1) return n;
  return fibRecursive(n - 1) + fibRecursive(n - 2);
}

// Memoization (Top-Down DP)
function fibMemo(n) {
  const memo = {};
  function fib(n) {
    if (n <= 1) return n;
    if (memo[n] !== undefined) return memo[n];
    return (memo[n] = fib(n - 1) + fib(n - 2));
  }
  return fib(n);
}

// Tabulation (Bottom-Up DP)
function fibTab(n) {
  if (n <= 1) return n;
  const dp = [0, 1];
  for (let i = 2; i <= n; i++) {
    dp[i] = dp[i - 1] + dp[i - 2];
  }
  return dp[n];
}

// Space Optimized
function fibSpace(n) {
  if (n <= 1) return n;
  let a = 0,
    b = 1;
  for (let i = 2; i <= n; i++) {
    [a, b] = [b, a + b];
  }
  return b;
}

File: dynamic_programming/01_Fibonacci/go/01_Fibonacci.go
package main

import (
	"fmt"
)

func main() {
	testCases := []struct {
		n      int
		expect int
	}{
		{0, 0},
		{1, 1},
		{2, 1},
		{3, 2},
		{4, 3},
		{5, 5},
		{10, 55},
		{20, 6765},
		{30, 832040},
	}

	fibFuncs := map[string]func(int) int{
		"Recursive":      fibRecursive,
		"Memoization":    fibMemoization,
		"Tabulation":     fibTabulation,
		"SpaceOptimized": fibSpaceOptimized,
	}

	for name, fibFunc := range fibFuncs {
		fmt.Printf("Testing %s:\n", name)
		for i, tc := range testCases {
			fmt.Printf("Test Case %d: fib(%d)\n", i+1, tc.n)
			res := fibFunc(tc.n)
			fmt.Printf("Got: %d\n", res)
			fmt.Printf("Expected: %d\n", tc.expect)
			if res != tc.expect {
				fmt.Println("FAIL")
			} else {
				fmt.Println("PASS")
			}
			fmt.Println()
		}
		fmt.Println("--------------------") // Separator between function tests
	}
}

// 1. Recursion
func fibRecursive(n int) int {
	if n <= 1 {
		return n
	}
	return fibRecursive(n-1) + fibRecursive(n-2)
}

// 2. Memoization
func fibMemoization(n int) int {
	memo := make(map[int]int)
	var fib func(n int) int
	fib = func(n int) int {
		if n <= 1 {
			return n
		}
		if val, ok := memo[n]; ok {
			return val
		}
		memo[n] = fib(n-1) + fib(n-2)
		return memo[n]
	}
	return fib(n)
}

// 3. Tabulation
func fibTabulation(n int) int {
	if n <= 1 {
		return n
	}
	dp := make([]int, n+1)
	dp[0] = 0
	dp[1] = 1
	for i := 2; i <= n; i++ {
		dp[i] = dp[i-1] + dp[i-2]
	}
	return dp[n]
}

// 4. Space Optimization
func fibSpaceOptimized(n int) int {
	if n <= 1 {
		return n
	}
	a, b := 0, 1
	for i := 2; i <= n; i++ {
		a, b = b, a+b
	}
	return b
}
